pages/learn/develop/container-contracts.md

```diff
@@ -0,0 +1,112 @@
+
+## Contract specification
+
+Container contracts are defined as a `contract.yml` file and must be placed in the root of the build context for a service. For example, the following [multi-container application][multicontainer] defines two services named `first-service` and `second-service`, each with a corresponding `contract.yml` file.
+
+```shell
+.
+├── docker-compose.yml
+├── first-service
+│ ├── contract.yml
+│ └── Dockerfile.template
+└── second-service
+  ├── contract.yml
+  └── Dockerfile.template
+```
+
+`contract.yml` is a YAML file containing `type`, `slug`, `name`, and `requires` keys, which define the requirements of the contract. The following example requires the device to be running a version of L4T equal to 28.2 for the contract to pass. For L4T exact versions (either 28.2 or 32.2) must be specified.
+
+```yaml
+type: "sw.container"
+slug: "enforce-l4t-version"
+name: "Enforce L4T Version"
+requires:
+  - type: "sw.l4t"
+    version: "28.2"
+```
+
+- `type` should be set to `sw.container` for container contracts.
+- `slug` should be a unique identifier to identify the contract.
+- `name` should be a descriptive name of the contract.
+- `requires` should be a list of requirements to be enforced by the contract defined by `type` and `version` keys. Currently, `type` may be either `sw.supervisor` for the device Supervisor version or `sw.l4t` for the L4T version.
+
+**Note** If `name`, `type`, or `slug` are omitted, the build will fail.
+
+Multiple requirements may be added to a single contract. The following example requires that the Supervisor version is greater than or equal to 10.6.13, and the L4T version is equal to 32.2.
+
+```yaml
+type: "sw.container"
+slug: "enforce-supervisor-and-l4t"
+name: "Enforce supervisor and l4t requirements"
+requires:
+  - type: "sw.supervisor"
+    version: ">=10.6.17"
+  - type: "sw.l4t"
+    version: "32.2"
+```
+
+**Note** Any type other than `sw.supervisor` or `sw.l4t` will result in the contract being invalid.
+
+When deploying a release, if the contract requirements are not met, the release will fail, and all devices in the application will remain on their current release. This default behavior can be overridden by including an [optional label](#optional-containers) in the application docker-compose.yml file. This label specifies which services with unmet requirements will be ignored, with the release being deployed to all other devices in the application. If there are any existing running containers for services with unmet requirements, they will not be destroyed and continue running the previous release.
+
+Should a contract fail, the dashboard logs will contain detail about the failing contract:
+
+```shell
+Could not move to new release: Some releases were rejected due to having unmet requirements:
+Contracts: Services with unmet requirements: first-service
```

Comment:
I wonder if it'd also be worth mentioning the contract slug in these error logs?